packages/ui-components/src/components/PopupMenu/PopupMenu.component.tsx

```diff
@@ -0,0 +1,394 @@
+  children?: React.ReactNode
+  /** Add a custom className to the wrapping element. NOTE: The Menu will be rendered into a Portal outside the wrapping parent element, so the Menu and its children will be outside the scope of the parent CSS selector.*/
+  className?: string
+  /** Whether the PopupMenu is disabled. */
+  disabled?: boolean
+  /** The icon to render when using the default toggle. Will be ignored if a PopupMenu.Toggle child is passed. */
+  icon?: keyof typeof KnownIconsEnum
+  /** The size of the menu and its items. */
+  menuSize?: "normal" | "small"
+  /** Handler to run when the Menu closes. */
+  onClose?: () => void
+  /** Handler to run when the Menu opens. */
+  onOpen?: () => void
+}
+
+export interface PopupMenuToggleProps extends React.ComponentProps<typeof HeadlessMenu.Button> {
+  className?: string
+}
+
+//Extract types directly from React.ElementType as Headless Menu Items is a dynamically typed component that cannot be extended with React.ComponentPropsWithRef (such as `href` or `target` – these will be accepted when passing `as` as `a` in order to render an anchor, but not when rendering a div.):
+type HeadlessMenuItemsProps = React.ComponentPropsWithRef<React.ElementType>
+// Extend the extracted props instead:
+export interface PopupMenuMenuProps extends HeadlessMenuItemsProps {
+  as?: React.ElementType // Allow customizing the element type just as headless ui does
+  className?: string
+  children?: React.ReactNode
+  key?: React.Key
+}
+
+export interface PopupMenuItemProps extends React.ComponentProps<typeof HeadlessMenu.Item> {
+  as?: React.ElementType // Allow customising the element type to allow for rendering items as anchor elements, just as headless ui does
+  className?: string
+  disabled?: boolean
+  href?: string // accept a href when rendering a link, whether or not to pass it to HUI item will behandled in the component
+  icon?: keyof typeof KnownIconsEnum
+  label?: string
+  rel?: string // accept rel for anchors, handle in component
+  target?: string // accepot target for links, handle in component
+}
+
+export interface PopupMenuSectionProps {
+  children?: React.ReactNode
+  className?: string
+}
+
+export interface PopupMenuSectionHeadingProps {
+  children?: React.ReactNode
+  className?: string
+  label?: string
+}
+
+export interface PopupMenuSectionSeparatorProps {
+  className?: string
+}
+
+// ----- Define PopupMenu context and hook -----
+
+// Context:
+const PopupMenuContext = createContext<PopupMenuContextType | null>(null)
+
+// Hook to use the context:
+export const usePopupMenuContext = () => {
+  const context = useContext(PopupMenuContext)
+  if (!context) {
+    throw new Error(
+      "The usePopupMenuContext hook must be used inside within a PopupMenu that provides the PopupMenuContext."
+    )
+  }
+  return context
+}
+
+// Export the context for advanced use cases:
+export { PopupMenuContext }
```

Comment:
Could you please clarify why do you need a separated exported context?

Reply:
Users can import the context in order to read the open state of the menu, so they can apply their own styling on custom toggle elements depending on the open state of the menu, or implement any other styling or funcitonality depending on its state. The context also exposes a function to close the menu, so any arbitrary element in or outside the menu can be configured to close it. `PopupMenu`is designed as a rather generic component, so the menu could contain elements other than `PopupMenu.Item`.

Reply:
p.s. I have updated the "With Toggle As Button Styled By State" story to better illustrate the concept and how to use the context.